osu.Game/Screens/Play/HUD/UnstableRateCounter.cs

```diff
@@ -0,0 +1,183 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using JetBrains.Annotations;
+using osu.Framework.Allocation;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Sprites;
+using osu.Framework.Localisation;
+using osu.Game.Beatmaps;
+using osu.Game.Graphics;
+using osu.Game.Graphics.Sprites;
+using osu.Game.Graphics.UserInterface;
+using osu.Game.Rulesets.Judgements;
+using osu.Game.Rulesets.Scoring;
+using osu.Game.Skinning;
+
+namespace osu.Game.Screens.Play.HUD
+{
+    public class UnstableRateCounter : RollingCounter<double>, ISkinnableDrawable
+    {
+        public bool UsesFixedAnchor { get; set; }
+
+        protected override double RollingDuration => 750;
+
+        private const float alpha_when_invalid = 0.3f;
+
+        private readonly List<double> hitOffsets = new List<double>();
+
+        //May be able to remove the CanBeNull as ScoreProcessor should exist everywhere, for example, in the skin editor it is cached.
+        [CanBeNull]
+        [Resolved(CanBeNull = true)]
+        private ScoreProcessor scoreProcessor { get; set; }
+
+        public UnstableRateCounter()
+        {
+            Current.Value = 0.0;
+        }
+
+        [BackgroundDependencyLoader]
+        private void load(OsuColour colours, BeatmapDifficultyCache difficultyCache)
+        {
+            Colour = colours.BlueLighter;
+        }
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            if (scoreProcessor == null) return;
+
+            scoreProcessor.NewJudgement += onJudgementAdded;
+            scoreProcessor.JudgementReverted += onJudgementReverted;
+        }
+
+        private bool isValid;
+
+        private void setValid(bool valid)
+        {
+            if (isValid == valid) return;
+
+            DrawableCount.FadeTo(valid ? 1 : alpha_when_invalid, 1000, Easing.OutQuint);
+            isValid = valid;
+        }
+
+        private void onJudgementAdded(JudgementResult judgement)
+        {
+            if (!(judgement.HitObject.HitWindows is HitWindows.EmptyHitWindows) && judgement.IsHit)
+            {
+                hitOffsets.Add(judgement.TimeOffset);
+            }
+
+            updateUr();
+        }
+
+        // If a judgement was reverted successfully, remove the item from the hitOffsets list.
+        private void onJudgementReverted(JudgementResult judgement)
+        {
+            //Score Processor Conditions to revert
+            if (judgement.FailedAtJudgement || !judgement.Type.IsScorable())
+                return;
+            //UR Conditions to Revert
+            if (judgement.HitObject.HitWindows is HitWindows.EmptyHitWindows || !judgement.IsHit)
+                return;
```

Comment:
not sure I like all these checks here - they look to be duplicating `ScoreProcessor` logic. would rather expose the `hitResults` list in `ScoreProcessor` as read-only.

Reply:
~~Wouldn't using the hitResults in scoreprocessor be more costly though? We would have to do something like~~
```cs
hitOffsets = scoreProcessor.hitEvents
    .Where(e => !(e.HitObject.HitWindows is HitWindows.EmptyHitWindows) && e.Result.IsHit())
    .Select(ev => ev.TimeOffset);
```
~~Instead of having the 4 checks.~~


While researching the code to get to discover this, I found that we could remove the `judgement.Type.IsScorable()` outright since the isHit property already covers that and more, leaving just the `judgement.FailedAtJudgement`.
I am unsure if you still want me to go through with the exposing hitResults/using gameplaystate's already exposed hitresults.

Reply:
I guess I could live with that last variant.